contracts/p1/BasketHandler.sol

```diff
@@ -361,70 +364,72 @@ contract BasketHandlerP1 is ComponentP1, IBasketHandler {
     // Otherwise returns (token's basket.refAmts / token's Collateral.refPerTok())
     function quantityUnsafe(IERC20 erc20, IAsset asset) public view returns (uint192) {
         if (!asset.isCollateral()) return FIX_ZERO;
-        return _quantity(erc20, ICollateral(address(asset)), CEIL);
+        return _quantity(erc20, ICollateral(address(asset)), false, CEIL);
     }
 
     /// @param erc20 The token contract
     /// @param coll The registered collateral plugin contract
-    /// @return {tok/BU} The token-quantity of an ERC20 token in the basket.
+    /// @param applyIssuancePremium Whether to apply an issuance premium to the quantity
+    /// @return q {tok/BU} The token-quantity of an ERC20 token in the basket
     // Returns 0 if coll is not in the basket
     // Returns FIX_MAX (in lieu of +infinity) if Collateral.refPerTok() is 0.
     // Otherwise returns (token's basket.refAmts / token's Collateral.refPerTok())
     function _quantity(
         IERC20 erc20,
         ICollateral coll,
+        bool applyIssuancePremium,
         RoundingMode rounding
-    ) internal view returns (uint192) {
+    ) internal view returns (uint192 q) {
         uint192 refPerTok = coll.refPerTok();
         if (refPerTok == 0) return FIX_MAX;
 
         // {tok/BU} = {ref/BU} / {ref/tok}
-        return basket.refAmts[erc20].div(refPerTok, rounding);
-    }
+        q = basket.refAmts[erc20].div(refPerTok, rounding);
 
-    /// Should not revert
-    /// @return low {UoA/BU} The lower end of the price estimate
-    /// @return high {UoA/BU} The upper end of the price estimate
-    // returns sum(quantity(erc20) * price(erc20) for erc20 in basket.erc20s)
-    function price() external view returns (uint192 low, uint192 high) {
-        return _price(false);
-    }
+        // Prevent toxic issuance by charging more when collateral is under peg
+        if (!skipIssuancePremium && applyIssuancePremium && coll.lastSave() == block.timestamp) {
+            // on arbitrum the timestamp check doesn't give us exactly what we want
+            // but it's close and better than wasting more gas on calling tryPrice()
 
-    /// Should not revert
-    /// lowLow should be nonzero when the asset might be worth selling
-    /// @dev Deprecated. Phased out in 3.1.0, but left on interface for backwards compatibility
-    /// @return lotLow {UoA/BU} The lower end of the lot price estimate
-    /// @return lotHigh {UoA/BU} The upper end of the lot price estimate
-    // returns sum(quantity(erc20) * lotPrice(erc20) for erc20 in basket.erc20s)
-    function lotPrice() external view returns (uint192 lotLow, uint192 lotHigh) {
-        return _price(true);
+            uint192 pegPrice = coll.savedPegPrice(); // {target/ref}
```

Comment:
This is a significant breaking change right, `savedPegPrice` doesn't exist in any existing collateral so all plugins MUST be upgraded to a new version when upgrading.

Reply:
Corrrrrect. I was initially thinking this would be okay because we'd just deploy 4.0.0 versions across the board, but with 3.4.0 there were a few cases of collateral left behind on <3.4.0 versions. 

Options seem to be try-catch or make sure to have 4.0.0 versions of everything.

Reply:
It's too bad this isn't a fit for the AssetPluginRegistry...it's no use for already-registered assets (on purpose)

Reply:
Went with try-catch